tests: fix these tests when using EFI console output function as default

enjiamai opened the pull request:
This PR try to fix some testcases when applying the UEFI console as default output （with UART_CONSOLE disable)

Category 1: Implemented the simplest driver to support EFI console:
- Add a minimal EFI console driver to support printf, fix following tests: 
tests/benchmarks/sys_kernel/ 
samples/application_development/external_lib/  
samples/posix/eventfd/
samples/userspace/hello_world_user/

- Implement a logging backend EFI console to fix following tests:  
samples/subsys/logging/syst
tests/subsys/logging/log_core_additional/
samples/subsys/logging/logger/

Category 2:  And there is a firmware bug that EFI console output call return with interrupt enabled. Though
there is a interrupt disabled solution that disable the interrupt right after EFI call, in some case
it the interrupt could happen during the EFI call context. If an interrupt was handled, a printk
call again will make it re-entries, or a swap might be happens.

- Not do prink in the interrupt context, and lock the scheduler during the EFI call.
samples/philosophers/
tests/kernel/interrupt/
tests/kernel/semaphore/sys_sem/
samples/kernel/metairq_dispatch/

- No tracing during switch in/out when using EFI console:
samples/subsys/tracing/

Fixes: #45929

enjiamai committed:
samples: subsys: tracing: No tracing during switch in/out using EFI console

Due to the limitation of the current x86 EFI cosnole implementation,
we skip hte printk which is called by tracing API:
sys_trace_thread_switched_in_user/out_user.

Signed-off-by: Enjia Mai <enjia.mai@intel.com>

enjiamai committed:
drivers: console: add a minimal EFI console driver to support printf

Add a minimal EFI console driver to support printf, this console driver
only supports console output. Otherwise the printf will not work.

Signed-off-by: Enjia Mai <enjia.mai@intel.com>

andyross commented on drivers/console/efi_console.c:
```diff
@@ -0,0 +1,85 @@
+#include <zephyr/sys/printk.h>
+
+
+#ifndef CONFIG_UART_CONSOLE
```
If you're putting an entire file under an #if, it's best to just filter it out of the build using cmake instead.

andyross reviewed (dismissed):
One nitpick, seems very reasonable.

cfriedt commented on subsys/tracing/user/tracing_user.c:
```diff
@@ -19,24 +19,30 @@ void __weak sys_trace_idle_user(void) {}
 {
+/* FIXME: Limitation of the current x86 EFI cosnole implementation. */
+#if !defined(CONFIG_X86_EFI_CONSOLE) && !defined(CONFIG_UART_CONSOLE)
+
```
I wonder if it might be more appropriate to use `IS_ENABLED()` here as well as in the function below.

enjiamai replied:
I think IS_ENABLED() is not target for this scenario, most use in the "if( IS_ENABLE(CONFIG_xxx))" code, thanks.

cfriedt commented on subsys/logging/Kconfig:
```diff
@@ -4,6 +4,7 @@
 menuconfig LOG
 	bool "Logging"
 	select PRINTK if USERSPACE
+	select UART_CONSOLE if X86_64
```
There may be some use cases where it makes sense to not use this option.

enjiamai replied:
Good point, I am still trying narrow this down.

enjiamai replied:
Got this fixed, thanks.

cfriedt commented on lib/libc/Kconfig:
```diff
@@ -221,7 +221,7 @@ endif # MINIMAL_LIBC
 config STDOUT_CONSOLE
 	bool "Send stdout to console"
-	depends on CONSOLE_HAS_DRIVER
+	depends on CONSOLE_HAS_DRIVER || X86_EFI_CONSOLE
```
Seems like it would be better to select `CONSOLE_HAS_DRIVER` when `X86_EFI_CONSOLE` is chosen.

enjiamai replied:
Yes, you are right. And I found the problem, we don't need this anymore. Thanks for pointing this out!

cfriedt commented on drivers/console/efi_console.c:
```diff
@@ -0,0 +1,82 @@
+
+extern int efi_console_putchar(int c);
+
+#if defined(CONFIG_PRINTK) || defined(CONFIG_STDOUT_CONSOLE)
```
I would suggest using
```cpp
if (IS_ENABLED(CONFIG_PRINTK) || defined(CONFIG_STDOUT_CONSOLE)
```
not 100% sure if this should be necessary though - is the conditional compile for `CONFIG_X86_EFI_CONSOLE` sufficient?

enjiamai replied:
Same as the top comment about IS_ENABLED macro. Thanks for your review.

enjiamai commented:
Hi @cfriedt , sorry for the late reply and think you very much for the review, could you please see my comments, thanks! :)

enjiamai commented:
Not passing the CI due to:
```
549/644 mps2_an385   tests/net/tcp/net.tcp.no_recv_queue   FAILED Failed (qemu 4.528s)
```

enjiamai committed:
logging: add a logging backend of EFI console

Add a EFI console logging backend to support logging sub-system.
To fix those testcases have no log output due to no logging backend.

Signed-off-by: Enjia Mai <enjia.mai@intel.com>

enjiamai committed:
arch: x86: workaround for EFI call return with interrupt enabled

The EFI console output call return with interrput enabled, it is a
firmware bug. And there was a solution that disabled interrupt it
return right away. But in some case the interrupt could happen
during the efi call context. If an interrupt was handled, a printk
call again will make it re-entried, or a swap might be happens.
This is suggested solution appiled for EFI console output:

1. Skip printk call when it is called in interrupt context.
2. Disable the schedule during the EFI call window.

Signed-off-by: Enjia Mai <enjia.mai@intel.com>

enjiamai commented:
Hi @andyross and dear all, I made a workaround solution for the EFI console with interrupt enabled problem, (https://github.com/zephyrproject-rtos/zephyr/pull/46918/commits/2bb56b792dd368db3206722c62f9ad3cdbe8ed5e).  Could you please give some suggestion or correct me if I am wrong, thank you!

And I also made a simplest t console driver and logging backend for EFI console. @cfriedt  :)

cfriedt reviewed (dismissed):
LGTM

andyross commented on subsys/logging/log_backend_efi_console.c:
```diff
@@ -0,0 +1,90 @@
+#include <zephyr/logging/log_backend_std.h>
+#include <zephyr/logging/log.h>
+
+LOG_MODULE_REGISTER(log_efi);
```
Not related to this PR, but I continue to be a little distressed at how much boilerplate is required for a log backend.

enjiamai replied:
Reduced the unnecessary include file. Thanks.

andyross reviewed (dismissed):
Looks great to me.

enjiamai commented:
Excuse me @andyross , @cfriedt , could you please help to re-approve / review again? thank you!
